### PR TITLE
Deduplicate declarations on insertion, not at the end of parsing a block

### LINKED ISSUE #15558: Hoist seen-property check before the virtual call and remove property deduplication
I did some digging to day with regards to @SimonSapin's question in #15531. I didn't have time to fully confirm my analysis with a profiler, but I think we should remove property deduplication unless there's a way to get it for free [1]. The |seen| check should mostly eliminate the cost of additional declarations in a block.

According to #15531, deduplication is primarily aimed at optimizing the cascade. But the cascade looks like this:
(1) Iterate all the declarations.
(2) For each declaration, invoke the appropriate virtual method.
(3) At the top of the virtual method, check the bitfield to see if we've seen the property. If we have, early-return.

(1) Involves walking the rule tree to find all the DeclarationBlocks. This is currently more expensive than it should be (see [2]), but eventually there should be only the minor cost of cache misses for pointer chasing. Regardless, deduplication doesn't reduce the number of DeclarationBlocks, but rather the number of declarations within a block. And those are stored in a Vec, so the cost of iterating a few extra declarations should be negligible.

(2) is a virtual call, which is where the current design really penalizes us for having additional declarations. But we can solve that by hoisting the |seen| check to the caller (this will give us a general performance benefits even ignoring the deduplication issue). If we do that, redundant declarations will get fast-rejected with negligible overhead.

Given all that, the only remaining benefit of deduplication is memory overhead. But given the price we're paying the CSS parser, I don't think it's worth it.

CC @emilio @Manishearth @rwakulszowa @heycam 

[1] According to @bzbarsky, Gecko uses a large stack-allocated property-keyed array at parse time, so new properties just overwrite the old ones - but then that array needs to be converted to a sparse representation somehow.
[2] https://bugzilla.mozilla.org/show_bug.cgi?id=1339703

### PR DESCRIPTION
**Pull request by SimonSapin**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15558 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15856)

### CODE REVIEW

**Comment by SimonSapin** [comment]
r? @Manishearth 
CC @bholley 
@bors-servo try

**Comment by metajack** [comment]
Trying http://help.appveyor.com/discussions/problems/4830-stuck-on-appveyor-was-unable-to-build-non-mergeable which suggests reopening may fix appveyor.

**Comment by SimonSapin** [comment]
Rebased.

@bors-servo try

**Review by Manishearth — APPROVED** [review]
r+ , probably should have @bholley check the perf

File: components/style/properties/declaration_block.rs [hunk 1]
```diff
@@ -169,19 +177,33 @@ impl PropertyDeclarationBlock {
...
+
+    /// Adds or overrides the declaration for a given property in this block,
+    /// Returns whether the declaration block is actually changed.
     pub fn set_parsed_declaration(&mut self,
                                   declaration: PropertyDeclaration,
                                   importance: Importance) -> bool {
+        self.push_common(declaration, importance, true)
+    }
+
+    fn push_common(&mut self, declaration: PropertyDeclaration, importance: Importance,
```
Comment (Manishearth) [inline components/style/properties/declaration_block.rs]: I feel like we could have more comments about this new design here, mentioning the word "deduplicate" at least once 😄 (for greppability)

**Pull request merged**